### PR TITLE
add functions to run dentist_detect_outliers (overall) and dentist_impute_single_window (single window)

### PR DESCRIPTION
**Pull request by RuiDongDR**
add functions to run `dentist_detect_outliers` (overall) and `dentist_impute_single_window` (single window)

```R
dentist_detect_outliers(sum_stat, LDmat, nSample = 1153)
```
and it will call (e.g., for window k)
```R
dentist_impute_single_window(zScore_k, LDmat_k, nSample)
```

There are more functions defined in this file to run the main function.

### CODE REVIEW

File: R/dentist_qc.R [hunk 1]
```diff
@@ -39,10 +39,14 @@
...
-                    pValueThreshold = 5e-8, propSVD = 0.4, gcControl = FALSE,
-                    nIter = 10, gPvalueThreshold = 0.05, ncpus = 1, seed = 999, 
-                    correct_chen_et_al_bug = TRUE) {
+dentist_impute_single_window <- function(zScore, LDmat, nSample,
+                                         pValueThreshold = 5e-8, propSVD = 0.4, gcControl = FALSE,
+                                         nIter = 10, gPvalueThreshold = 0.05, ncpus = 1, seed = 999, correct_chen_et_al_bug = TRUE) {
+  # Check that number of variants cannot be below 2000
+  if (length(zScore) < 2000) {
+    warning("The number of variants is below 2000. The function will not be executed.")
+    return(NULL) # Return NULL to indicate no results
```
Comment (gaow) [inline R/dentist_qc.R:48]: Let's not do this. Just a warning is enough. Let's say that the number is < 2000 the algorithm may not work as expected, as suggested by the DENTIST authors.

File: R/dentist_qc.R [hunk 2]
```diff
@@ -70,11 +73,180 @@ dentist <- function(zScore, LDmat, nSample,
...
+dentist_detect_outliers <- function(sum_stat, LDmat, nSample,
+                                    window_size = 2000000, pValueThreshold = 5e-8, propSVD = 0.4, gcControl = FALSE,
+                                    nIter = 10, gPvalueThreshold = 0.05, ncpus = 1, seed = 999, correct_chen_et_al_bug = TRUE) {
+  # detect for column names and order by pos
+  if (!any(tolower(c("pos", "position")) %in% tolower(colnames(sum_stat))) ||
+    !any(tolower(c("z", "zscore")) %in% tolower(colnames(sum_stat)))) {
+    stop("Input sum_stat is missing either 'pos'/'position' or 'z'/'zscore' column.")
+  }
+  sum_stat <- sum_stat %>% arrange(pos)
+  ### FIXME: formalize this logic somewhere
```
Comment (gaow) [inline R/dentist_qc.R:207]: You can use AI to help you integrate this to the roxygen2 documentation string. You don't have that yet anyways. Just copy your entire function to one of the AI chats asking it to convert the FIXME comment, along with the interface of the function, into roxygen2 documentation. And you add it here.

File: R/dentist_qc.R [hunk 3]
```diff
@@ -70,11 +73,180 @@ dentist <- function(zScore, LDmat, nSample,
...
+  ### if correct_chen_et_al_bug = FALSE, then there can never be a single window
+  # 1. if window_size is invalid (<=0 or >=range), and
+  # 1.1. if correct_chen_et_al_bug==TRUE, run dentist single window, so just window (0-4122)
+  # 1.2. if correct_chen_et_al_bug==FALSE, run in the dentist original way, so window (0-4122) and (742-4122)
+  # 2. if window_size is valid, and
+  # 2.1 if correct_chen_et_al_bug==TRUE, divide the window but don't add additional window if the first window covers all, so just (0-4122)
+  # 2.2 if correct_chen_et_al_bug==FALSE, divide the window in the dentist original way, so window (0-4122) and (742-4122)
+  if (window_size <= 0 | ((window_size >= max(sum_stat$pos) - min(sum_stat$pos) | is.na(window_size)) & (correct_chen_et_al_bug == TRUE))) {
+    print("single window!")
```
Comment (gaow) [inline R/dentist_qc.R:216]: Dont print. It is messy

File: R/dentist_qc.R [hunk 4]
```diff
@@ -70,11 +73,180 @@ dentist <- function(zScore, LDmat, nSample,
...
+    imputed_result <- dentist_impute_single_window(
+      sum_stat$z, LDmat, nSample,
+      pValueThreshold, propSVD, gcControl,
+      nIter, gPvalueThreshold, ncpus, seed, correct_chen_et_al_bug
+    )
+  } else {
+    # divide windows
+    window_divided_res <- divide_into_windows(sum_stat$pos, window_size = window_size, correct_chen_et_al_bug = TRUE)
+    print(window_divided_res)
```
Comment (gaow) [inline R/dentist_qc.R:225]: Dont print. It is messy

File: R/dentist_qc.R [hunk 5]
```diff
@@ -70,11 +73,180 @@ dentist <- function(zScore, LDmat, nSample,
...
+      )
+    }
+    # merge imputed result and generate a final imputed_result (similar to imputed_result above)
+    imputed_result <- merge_windows(imputed_result_by_window, window_divided_res)
+  }
+  # detect outlier
+  lambda_original <- 1
+  imputed_result <- imputed_result %>%
+    mutate(
+      stat = calculate_stat(original_z, imputed_z, rsq),
```
Comment (gaow) [inline R/dentist_qc.R:247]: Please rename this -- from `stat` to `outlier_stat`

File: R/dentist_qc.R (hunk 1 above)
Comment (RuiDongDR) [reply R/dentist_qc.R]: removed NULL and fixed the message.

File: R/dentist_qc.R (hunk 4 above)
Comment (RuiDongDR) [reply R/dentist_qc.R]: removed.

File: R/dentist_qc.R (hunk 5 above)
Comment (RuiDongDR) [reply R/dentist_qc.R]: renamed.

**Comment by gaow** [comment]
@RuiDongDR good work!

**Pull request merged**